utils/reusable_buffer.hh

```
@@ -9,13 +9,27 @@
 #pragma once
 
 #include "utils/assert.hh"
+#include "utils/exception_container.hh"
 #include "utils/fragmented_temporary_buffer.hh"
+#include "utils/result.hh"
 #include <seastar/core/timer.hh>
 #include <seastar/core/memory.hh>
 #include <bit>
+#include <concepts>
 
 namespace utils {
 
+template<typename Fn, typename Ret>
+concept buffer_writer_for = std::invocable<Fn, bytes_mutable_view>
+    && ExceptionContainerResult<std::invoke_result_t<Fn, bytes_mutable_view>>
+    && std::same_as<typename std::invoke_result_t<Fn, bytes_mutable_view>::value_type, size_t>;
+
+template<typename Fn>
+concept bytes_ostream_writer = buffer_writer_for<Fn, bytes_ostream>;
+
+template<typename Fn>
+concept fragmented_buffer_writer = buffer_writer_for<Fn, fragmented_temporary_buffer>;
+
```

Why these changes?

---

`reusable_buffer` needs functions with specific signature. While these concepts are not technically required, and what they achieve could be achieved in another way, they do make the compiler fail early and with clearer error messages in case an incompatible function is passed.

---

The patch does not introduce `reusable_buffer`. It is already used inside `transport/server.cc`.

This patch just changes the lambda signature from returning a value to returning result with either a value or an exception. This is done because we don't want to throw, and the previous lambda version that is passed to the `reusable_buffer` methods either returned a value or threw an exception.